# 1) Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px
import seaborn as sns

# 2) Load Clustered Data Frame

In [2]:
df=pd.read_csv("/workspaces/Customer_Segmentation_KMeans_Clustering/data/clusters.csv")
df.head()

,CustomerID,TotalSpent,Orders,ItemsPurchased,CountriesPurchased,AvgOrderValue,Recency,Cluster
0,12347.0,4310.00,7,2458,1,615.714286,1,2
1,12348.0,1797.24,4,2341,1,449.310000,74,0
2,12349.0,1757.55,1,631,1,1757.550000,18,0
3,12350.0,334.40,1,197,1,334.400000,309,1
4,12352.0,1240.73,4,380,1,310.182500,35,3


# 3) Cluster Analysis

## a) Analysis by Total Spent

In [3]:
total_spent_summary = (
    df
    .groupby("Cluster")["TotalSpent"]
    .agg(["mean", "median", "min", "max", "count"])
    .round(2)
)

total_spent_summary

,mean,median,min,max,count
Cluster,,,,,
0,1782.40,1396.61,419.58,39619.50,659
1,324.66,295.24,38.10,1661.06,722
2,5745.16,2862.11,601.56,259657.30,827
3,613.41,568.24,168.12,1759.50,1441
4,151.50,147.28,0.85,464.90,590


In [5]:


cluster_means = (
    df
    .groupby("Cluster", as_index=False)["TotalSpent"]
    .mean()
)

fig = px.bar(
    cluster_means,
    x="Cluster",
    y="TotalSpent",
    text_auto=".2f",
    title="Average Total Spending by Customer Cluster"
)

fig.update_layout(
    xaxis_title="Customer Cluster",
    yaxis_title="Average Total Spent ($)"
)

fig.show()

### Insights from Total Spent by Cluster

The average total spending varied considerably across the five clusters, indicating that the K-Means algorithm successfully separated customers based on their overall monetary value.

Cluster 2 – High-Value Customers (VIP Segment)
Mean Total Spent: $5,745.16 (Highest)
Median: $2,862.11

This cluster represents the highest-value customers. They contribute the greatest revenue to the business and include the customer with the highest recorded spending ($259,657.30). The large gap between the mean and median indicates the presence of a few exceptionally high spenders within this group. These customers should be prioritized for retention through loyalty programs, exclusive offers, premium customer service, and personalized marketing campaigns.

Cluster 0 – Above-Average Customers
Mean Total Spent: $1,782.40
Median: $1,396.61

Customers in this cluster spend substantially more than the overall average customer but considerably less than those in Cluster 2. The difference between the mean and median suggests moderate variation in spending, with a few higher-spending customers influencing the average. These customers represent an excellent target for upselling and cross-selling strategies aimed at increasing their lifetime value.

Cluster 3 – Moderate-Value Customers
Mean Total Spent: $613.41
Median: $568.24

This is the largest customer segment, comprising 1,441 customers. Spending levels are relatively consistent, as indicated by the small difference between the mean and median. These customers represent the core customer base and provide stable revenue. Promotional campaigns, product recommendations, and bundle offers could encourage them to increase their purchase value.

Cluster 1 – Low-Value Customers
Mean Total Spent: $324.66
Median: $295.24

Customers in this segment have relatively low spending levels. Since the mean and median are close, spending behaviour appears fairly consistent across the cluster. Marketing efforts should focus on increasing engagement through discounts, personalized recommendations, and incentives that encourage repeat purchases.

Cluster 4 – Very Low-Value Customers
Mean Total Spent: $151.50 (Lowest)
Median: $147.28

This cluster contains the lowest-spending customers in the dataset. The close agreement between the mean and median indicates consistently low spending behaviour. These customers may be new shoppers, occasional buyers, or customers with limited purchasing needs. Entry-level promotions, welcome offers, and targeted campaigns designed to encourage a second or third purchase may help increase their long-term value.

Overall Takeaway

The Total Spent analysis reveals a clear hierarchy of customer value across the five clusters. Cluster 2 contains the business's highest-value customers and should be the primary focus of retention strategies, while Cluster 0 consists of promising high-value customers with potential for further growth. Cluster 3 forms the largest and most stable customer base, making it an important target for campaigns aimed at increasing average spending. Finally, Clusters 1 and 4 represent lower-value customers who could benefit from acquisition, activation, and re-engagement campaigns designed to improve their purchasing behaviour.

## b) Analysis by Orders

In [6]:
orders_summary = (
    df
    .groupby("Cluster")["Orders"]
    .agg(["mean", "median", "min", "max", "count"])
    .round(2)
)

orders_summary

,mean,median,min,max,count
Cluster,,,,,
0,2.41,2.0,1,6,659
1,1.28,1.0,1,5,722
2,10.99,8.0,4,182,827
3,2.61,2.0,1,9,1441
4,1.42,1.0,1,6,590


In [7]:
cluster_orders = (
    df
    .groupby("Cluster", as_index=False)["Orders"]
    .mean()
)

fig = px.bar(
    cluster_orders,
    x="Cluster",
    y="Orders",
    text_auto=".2f",
    title="Average Number of Orders by Customer Cluster"
)

fig.update_layout(
    xaxis_title="Customer Cluster",
    yaxis_title="Average Number of Orders"
)

fig.show()

### Insights from Orders by Cluster

The average number of orders varied substantially across the five clusters, indicating distinct purchasing frequencies among the customer segments.

Cluster 2 – Highly Loyal Customers
Mean Orders: 10.99 (Highest)
Median: 8 orders

Cluster 2 stands out as the most loyal customer segment, placing an average of approximately 11 orders, which is significantly higher than all other clusters. The maximum of 182 orders suggests the presence of a few exceptionally active customers, but the relatively high median of 8 orders confirms that frequent purchasing is characteristic of the cluster as a whole. Combined with the previous finding that this cluster also has the highest total spending, these customers represent the business's most valuable and engaged customer base.

Cluster 3 – Regular Customers
Mean Orders: 2.61
Median: 2 orders

This is the largest customer segment and represents regular shoppers who make repeat purchases, although much less frequently than Cluster 2. Their purchasing behaviour is relatively consistent, making them a stable source of revenue. These customers may be encouraged to purchase more frequently through personalized promotions or loyalty incentives.

Cluster 0 – Occasional Customers
Mean Orders: 2.41
Median: 2 orders

Customers in this cluster purchase slightly less frequently than those in Cluster 3. However, from the previous analysis, they spend considerably more overall. This suggests that they tend to make larger purchases per shopping trip rather than purchasing frequently. Such customers may respond well to campaigns encouraging more frequent visits or repeat purchases.

Cluster 4 – Infrequent Customers
Mean Orders: 1.42
Median: 1 order

Most customers in this cluster have made only a single purchase. Their low purchasing frequency, together with their low total spending, indicates limited engagement with the business. These customers may benefit from re-engagement campaigns, first-repeat purchase discounts, or reminder emails.

Cluster 1 – One-Time or Low-Frequency Customers
Mean Orders: 1.28 (Lowest)
Median: 1 order

Cluster 1 contains customers with the lowest purchasing frequency. Most have placed only one order, suggesting they are either new customers or customers who have not returned after their initial purchase. Retention campaigns designed to encourage a second purchase would be appropriate for this group.

## c) Analysis by Items Purchased

In [8]:
items_summary = (
    df
    .groupby("Cluster")["ItemsPurchased"]
    .agg(["mean", "median", "min", "max", "count"])
    .round(2)
)

items_summary

,mean,median,min,max,count
Cluster,,,,,
0,1007.80,804.0,1,7824,659
1,169.36,139.5,2,1109,722
2,3372.88,1749.0,282,153567,827
3,369.04,316.0,17,2019,1441
4,87.28,76.0,1,574,590


In [9]:
cluster_items = (
    df
    .groupby("Cluster", as_index=False)["ItemsPurchased"]
    .mean()
)

fig = px.bar(
    cluster_items,
    x="Cluster",
    y="ItemsPurchased",
    text_auto=".2f",
    title="Average Number of Items Purchased by Customer Cluster"
)

fig.update_layout(
    xaxis_title="Customer Cluster",
    yaxis_title="Average Items Purchased"
)

fig.show()

### Insights from Items Purchased by Cluster

The analysis of ItemsPurchased indicates substantial variation in the quantity of products purchased across the five customer segments. This suggests that K-Means successfully distinguished customers based on their purchasing volume.

Cluster 2 – High-Volume Customers
Mean Items Purchased: 3,372.88 (Highest)
Median: 1,749 items

Cluster 2 purchased by far the largest number of items, reinforcing the findings from the previous analyses that this is the business's most valuable customer segment. The exceptionally high maximum value (153,567 items) indicates the presence of some bulk purchasers or wholesale customers. However, the high median demonstrates that purchasing large quantities is typical of this cluster rather than being driven solely by a few extreme customers.

Cluster 0 – Moderate-to-High Volume Customers
Mean Items Purchased: 1,007.80
Median: 804 items

Customers in Cluster 0 purchase a relatively high number of items despite placing only a moderate number of orders. Combined with their high total spending, this suggests they tend to make larger purchases per transaction. These customers appear to prefer fewer shopping trips while buying many products each time.

Cluster 3 – Moderate-Volume Customers
Mean Items Purchased: 369.04
Median: 316 items

As the largest customer segment, Cluster 3 exhibits moderate purchasing volumes. Their behaviour is relatively consistent, as shown by the small difference between the mean and median. These customers form the core customer base and offer opportunities for incremental growth through cross-selling and personalized recommendations.

Cluster 1 – Low-Volume Customers
Mean Items Purchased: 169.36
Median: 139.5 items

Customers in this cluster purchase relatively few items and, as seen previously, also spend relatively little and place few orders. This indicates a low level of engagement with the business. Marketing efforts should focus on encouraging repeat purchases and increasing basket size.

Cluster 4 – Very Low-Volume Customers
Mean Items Purchased: 87.28 (Lowest)
Median: 76 items

Cluster 4 represents customers with the lowest purchasing volume. Their consistently low spending, low purchase frequency, and small number of items suggest they are occasional or one-time buyers. These customers are ideal candidates for introductory promotions or campaigns aimed at increasing engagement.

## d) Analysis by Average Order Value

In [10]:
aov_summary = (
    df
    .groupby("Cluster")["AvgOrderValue"]
    .agg(["mean", "median", "min", "max", "count"])
    .round(2)
)

aov_summary

,mean,median,min,max,count
Cluster,,,,,
0,837.00,620.13,279.17,19809.75,659
1,266.36,233.33,35.15,1141.24,722
2,449.85,352.48,33.24,4525.16,827
3,264.31,250.67,51.12,739.20,1441
4,116.38,111.33,0.85,464.90,590


In [12]:
cluster_aov = (
    df
    .groupby("Cluster", as_index=False)["AvgOrderValue"]
    .mean()
)

fig = px.bar(
    cluster_aov,
    x="Cluster",
    y="AvgOrderValue",
    text_auto=".2f",
    title="Average Order Value by Customer Cluster"
)

fig.update_layout(
    xaxis_title="Customer Cluster",
    yaxis_title="Average Order Value ($)"
)

fig.show()

### Insights from Average Order Value by Cluster

The Average Order Value (AOV) analysis shows notable differences in spending behaviour across the five customer segments, indicating that customers generate revenue through different purchasing patterns.

Cluster 0 – Premium Buyers
Mean AOV: $837.00 (Highest)
Median: $620.13

Cluster 0 has the highest average order value of all customer segments, nearly double that of the next highest cluster. This indicates that although these customers place only a moderate number of orders, they spend a substantial amount during each purchase. Combined with the previous analyses, this suggests that Cluster 0 consists of premium or bulk buyers who prefer fewer but significantly larger transactions. These customers are ideal candidates for premium product recommendations, exclusive collections, and high-value product bundles.

Cluster 2 – Loyal High-Value Customers
Mean AOV: $449.85
Median: $352.48

Although Cluster 2 does not have the highest average order value, it combines a relatively high AOV with the highest purchase frequency and the highest total spending. This indicates that these customers generate value primarily through frequent repeat purchases rather than exceptionally expensive individual orders. Their consistent engagement makes them excellent candidates for loyalty rewards, personalized recommendations, and VIP membership programs.

Cluster 1 – Moderate Basket Customers
Mean AOV: $266.36
Median: $233.33

Customers in Cluster 1 have moderate spending per order but relatively low purchase frequency and total spending. Increasing their shopping frequency may be more beneficial than attempting to increase their basket size.

Cluster 3 – Regular Everyday Customers
Mean AOV: $264.31
Median: $250.67

Cluster 3 exhibits a very similar average order value to Cluster 1 but differs in purchasing behaviour by placing more orders and buying more items overall. This suggests these customers are regular shoppers who make consistent purchases over time. Their stable buying habits make them suitable targets for cross-selling, personalized product recommendations, and loyalty incentives designed to increase purchase frequency.

Cluster 4 – Budget Customers
Mean AOV: $116.38 (Lowest)
Median: $111.33

Customers in Cluster 4 have the lowest average order value, indicating that they make relatively small purchases whenever they shop. Combined with their low total spending and low purchase frequency, these customers represent the lowest-value customer segment. Marketing efforts should focus on increasing both basket size and repeat purchasing through entry-level promotions and product bundles.

## e) Analysis by Recency

In [13]:
recency_summary = (
    df
    .groupby("Cluster")["Recency"]
    .agg(["mean", "median", "min", "max", "count"])
    .round(2)
)

recency_summary

,mean,median,min,max,count
Cluster,,,,,
0,82.21,63.0,0,371,659
1,271.47,269.0,132,373,722
2,22.56,14.0,0,275,827
3,49.67,36.0,0,212,1441
4,90.54,64.0,0,371,590


In [14]:
cluster_recency = (
    df
    .groupby("Cluster", as_index=False)["Recency"]
    .mean()
)

fig = px.bar(
    cluster_recency,
    x="Cluster",
    y="Recency",
    text_auto=".2f",
    title="Average Recency by Customer Cluster"
)

fig.update_layout(
    xaxis_title="Customer Cluster",
    yaxis_title="Average Days Since Last Purchase"
)

fig.show()

## Insights from Recency by Cluster

The Recency analysis reveals clear differences in customer activity across the five clusters. Since Recency measures the number of days since a customer's last purchase, lower values indicate more recently active customers, while higher values indicate customers who have not purchased for a longer period.

Cluster 2 – Highly Active Customers
Mean Recency: 22.56 days (Lowest)
Median: 14 days

Cluster 2 consists of the most active customers, having made purchases within the past few weeks on average. Combined with their high spending, high purchase frequency, and high purchasing volume, these customers represent the retailer's most valuable and engaged customer segment. Retaining these customers should be a top priority through loyalty rewards, VIP programs, and personalized product recommendations.

Cluster 3 – Active Regular Customers
Mean Recency: 49.67 days
Median: 36 days

Customers in Cluster 3 have also purchased relatively recently, indicating ongoing engagement with the business. Although they spend less than Cluster 2, they represent the largest customer segment and provide consistent revenue. These customers are ideal candidates for cross-selling, personalized recommendations, and campaigns that encourage slightly more frequent purchases.

Cluster 0 – Moderately Active Customers
Mean Recency: 82.21 days
Median: 63 days

Cluster 0 has not purchased as recently as Clusters 2 and 3 but remains moderately active. Given their high average order value and relatively high total spending, these customers are valuable but may benefit from reminder emails, exclusive promotions, or new product announcements to encourage more frequent purchases.

Cluster 4 – At-Risk Customers
Mean Recency: 90.54 days
Median: 64 days

Customers in Cluster 4 have relatively long periods since their last purchase while also exhibiting the lowest spending and purchasing volume. These customers appear to be drifting away from the business and may require re-engagement campaigns such as discounts, personalized offers, or incentives to return.

Cluster 1 – Dormant Customers
Mean Recency: 271.47 days (Highest)
Median: 269 days

Cluster 1 has the highest recency by a wide margin, indicating that these customers have not made purchases for many months. Combined with their low spending and low purchase frequency, they represent the least engaged customer segment. These customers are at high risk of churn and should be targeted with win-back campaigns, special discounts, or surveys to understand why they have stopped purchasing.

## Full Clusters summary

In [15]:
cluster_summary = (
    df
    .groupby("Cluster")
    .mean(numeric_only=True)
    .reset_index()
)




# Reshape the dataframe
cluster_long = cluster_summary.melt(
    id_vars="Cluster",
    value_vars=[
        "TotalSpent",
        "Orders",
        "ItemsPurchased",
        "AvgOrderValue",
        "Recency"
    ],
    var_name="Feature",
    value_name="Value"
)

fig = px.bar(
    cluster_long,
    x="Cluster",
    y="Value",
    color="Feature",
    barmode="group",
    title="Average Feature Values by Customer Cluster"
)

fig.update_layout(
    xaxis_title="Cluster",
    yaxis_title="Average Value"
)

fig.show()

## Final Customer Profiles

| Cluster | Customer Profile                | Key Characteristics                                                                       | Recommended Marketing Campaign                                                                                         |
| :-----: | ------------------------------- | ----------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------- |
|  **0**  | **Premium / Bulk Buyers**       | High spending, high average order value, moderate purchase frequency, moderately active   | Premium product launches, exclusive bundles, personalized recommendations, reminders to increase purchase frequency.   |
|  **1**  | **Dormant Low-Value Customers** | Low spending, few orders, low purchasing volume, longest time since last purchase         | Win-back campaigns, reactivation emails, discount vouchers, limited-time offers, customer feedback surveys.            |
|  **2**  | **VIP Loyal Customers**         | Highest spending, most orders, largest purchasing volume, most recently active            | Loyalty rewards, VIP memberships, early access to new products, personalized offers, premium customer service.         |
|  **3**  | **Regular Repeat Customers**    | Moderate spending, moderate purchase frequency, active shoppers, largest customer segment | Cross-selling, upselling, loyalty points, product recommendations, bundle offers.                                      |
|  **4**  | **Occasional Budget Customers** | Lowest spending, lowest purchasing volume, infrequent purchases, becoming inactive        | Welcome-back promotions, first-repeat purchase discounts, entry-level product recommendations, price-based promotions. |


# Final Project Takeaway



The K-Means clustering model successfully segmented customers into five distinct groups, each exhibiting unique purchasing behaviours. The analysis revealed clear differences in customer value, purchasing frequency, buying volume, spending per transaction, and recent activity.

The most valuable segment (Cluster 2) consists of highly engaged customers who purchase frequently, spend the most overall, and remain actively engaged with the business. In contrast, Cluster 1 represents dormant customers who have not made purchases for an extended period and require re-engagement efforts. Cluster 0 contains premium buyers who make fewer but higher-value purchases, Cluster 3 forms the retailer's stable core customer base, and Cluster 4 consists of low-value, occasional shoppers with growth potential.

These findings demonstrate how unsupervised learning can transform transactional data into actionable business intelligence, enabling organizations to design targeted marketing campaigns, improve customer retention, optimize promotional strategies, and allocate marketing resources more effectively.

# 4) Proposed Marketing Campaign Strategy

The customer segmentation analysis identified five distinct customer groups, each exhibiting unique purchasing behaviours. Rather than applying a single marketing strategy to all customers, the business can leverage these customer segments to deliver personalized marketing campaigns that address the specific characteristics of each group. Such targeted marketing strategies can improve customer retention, increase sales, and optimize marketing expenditure.

| Cluster       | Customer Segment                | Key Characteristics                                                                           | Proposed Marketing Strategy                                                                                                                                                                                  |
| ------------- | ------------------------------- | --------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| **Cluster 0** | **Premium / Bulk Buyers**       | High total spending, highest average order value, moderate purchase frequency                 | Promote premium products, exclusive collections, bulk purchase discounts, and personalized product recommendations. Encourage more frequent purchases through exclusive offers and free shipping incentives. |
| **Cluster 1** | **Dormant Low-Value Customers** | Low spending, few orders, longest period since last purchase                                  | Launch win-back campaigns using discount vouchers, limited-time offers, reminder emails, and customer satisfaction surveys to re-engage inactive customers.                                                  |
| **Cluster 2** | **VIP Loyal Customers**         | Highest spending, highest purchase frequency, largest purchasing volume, most recently active | Reward customer loyalty through VIP memberships, loyalty points, early access to new products, personalized recommendations, exclusive promotions, and premium customer support to maximize retention.       |
| **Cluster 3** | **Regular Repeat Customers**    | Moderate spending, moderate purchasing frequency, active customer base                        | Increase customer lifetime value through cross-selling, upselling, product bundles, personalized recommendations, and loyalty programs that encourage more frequent purchases.                               |
| **Cluster 4** | **Occasional Budget Customers** | Lowest spending, lowest purchasing volume, relatively infrequent purchases                    | Encourage repeat purchases using welcome-back promotions, first-repeat purchase discounts, seasonal offers, and affordable product recommendations aimed at increasing basket size.                          |


#### Business Value

Implementing customer-specific marketing campaigns enables businesses to allocate marketing resources more efficiently by targeting customers according to their purchasing behaviour. High-value customers can be retained through loyalty and premium service initiatives, while lower-value and inactive customers can be re-engaged using promotional incentives. This personalized approach is expected to improve customer retention, increase customer lifetime value, enhance campaign effectiveness, and ultimately drive higher business revenue.

#### Practical Implementation

Although the dataset used in this project does not contain customer contact information such as names, email addresses, or phone numbers, these marketing strategies can be implemented in a real-world environment by integrating the generated customer segments with a Customer Relationship Management (CRM) system using the CustomerID field. This integration would allow the business to automatically assign customers to the appropriate marketing campaign and monitor campaign performance through metrics such as conversion rate, repeat purchases, and return on investment (ROI).

> This project demonstrates how unsupervised machine learning can transform raw transactional data into actionable customer insights. By identifying distinct customer segments and proposing targeted marketing strategies, the analysis provides a practical framework for improving customer engagement, optimizing marketing efforts, and supporting data-driven business decision-making.